pmd-core/src/main/java/net/sourceforge/pmd/benchmark/TimeTracker.java

```diff
@@ -0,0 +1,255 @@
+ * 
+ * @author Juan Martín Sotuyo Dodero
+ */
+public final class TimeTracker {
+
+    private static boolean trackTime = false;
+    private static long wallClockStart = -1;
+    private static final ThreadLocal<Queue<TimerEntry>> TIMER_ENTRIES;
+    private static final ConcurrentMap<TimedOperation, TimedResult> ACCUMULATED_RESULTS = new ConcurrentHashMap<>();
+    
+    static {
+        TIMER_ENTRIES = new ThreadLocal<Queue<TimerEntry>>() {
+            @Override
+            protected Queue<TimerEntry> initialValue() {
+                final Queue<TimerEntry> queue = Collections.asLifoQueue(new LinkedList<TimerEntry>());
+                queue.add(new TimerEntry(TimedOperationCategory.UNACCOUNTED, null));
+                return queue;
+            }
+        };
+    }
+    
+    private TimeTracker() {
+        throw new AssertionError("Can't instantiate utility class");
+    }
+    
+    /**
+     * Starts global tracking. Allows tracking operations to take place and starts the wall clock.
+     * Must be called once PMD starts if tracking is desired, no tracking will be performed otherwise.
+     */
+    public static void startGlobalTracking() {
+        wallClockStart = System.currentTimeMillis();
+        trackTime = true;
+        initThread(); // init main thread
+    }
+    
+    /**
+     * Stops global tracking. Stops the wall clock. All further operations will be treated as NOOP.
+     * @return The timed data obtained through the run.
+     */
+    public static TimingReport stopGlobalTracking() {
+        if (!trackTime) {
+            return null;
+        }
+        
+        finishThread(); // finish the main thread
+        trackTime = false;
+        
+        // Fix UNACCOUNTED metric (total time is meaningless)
+        final TimedResult unaccountedResult = ACCUMULATED_RESULTS.get(
+                new TimedOperation(TimedOperationCategory.UNACCOUNTED, null));
+        unaccountedResult.totalTime.set(unaccountedResult.selfTime.get());
+        
+        return new TimingReport(System.currentTimeMillis() - wallClockStart, ACCUMULATED_RESULTS);
+    }
+    
+    /**
+     * Initialize a thread, starting to track it's own time.
+     */
+    public static void initThread() {
+        if (!trackTime) {
+            return;
+        }
+        
+        TIMER_ENTRIES.get(); // Just make sure it's initialized
+    }
+    
+    /**
+     * Finishes tracking a thread.
+     */
+    public static void finishThread() {
+        if (!trackTime) {
+            return;
+        }
+        
+        // if using a mono-thread, we may not be empty...
+        if (TIMER_ENTRIES.get().size() == 1) {
+            finishOperation();
+            TIMER_ENTRIES.remove();
+        }
+    }
+    
+    /**
+     * Starts tracking an operation.
+     * @param category The category under which to track the operation.
+     */
+    public static void startOperation(final TimedOperationCategory category) {
```

Comment:
I recall experimenting with refactoring to use an `AutoClosable` so that [try-with-resources](https://docs.oracle.com/javase/specs/jls/se7/html/jls-14.html#jls-14.20.3) could be used.  I think I was trying to fix bookkeeping in the face of internal processing exceptions while working on the regression report.  IIRC exceptions caused time to **_not_** get logged properly.  Internally slow processing because of exception throwing code misbehaving was not obvious in the benchmark report.  Problem isolation was harder because I couldn't tell which Rule was eating up all the time.  In your current bookkeeping, it would be like seeing 100s "wall", and only 10s "measured".  Where'd all the "measured" time go?!?!  😄  (Note: Using "measured" here not "cpu", seams like a misnomer.  Perhaps "thread" would be better, if "thread seconds" is meaningful?  Sorry digression...)

So, roughly in your current `TimeTracker` lingo:

```
// A means to both close() when done, and optionally while not closed accumulate via accept().
interface Tracker extends AutoClosable, LongConsumer {
   /** Add Javadoc here... */
   // Note: Overridden to NOT throw IOException
   @Override
   void close();

   /** Add Javadoc here... */
   void accept(long value);
}

// Meanwhile in some PMD code...

try (Tracker tracker = TimeTracker.startOperation(TimedOperationCategory.SOME_CATEOGRY)) {
   // Timed code goes inside here.  Will automatically call close() even when exceptions happen!

   // For example...
   int total = o;
   for (Object obj : objects) {
      // Do something interesting, then...

      tracker.accept(1); // Count by 1 in a loop
      total++;
   }
   tracker.accept(total); // Or use a total if available
}
```

I hope this makes sense.

Finally, one last bit, you could _remove_ use of `static` methods in normal PMD code, and just let call `new Tracker(TimedOperationCategory.XXX)` and have the implementation do all the internal routing to a static accumulator, etc.  It's really more a style choice to avoid calling into `static` global methods, and be more OO-like.

Hmm, independent of the `static` style, perhaps you could eliminate thread-locals, since a `Tracker` stores on the stack/heap naturally?

Reply:
I thought of something along those lines early on, but discarded it since I thought it would never be a nop, but reading your draft I realize I can actually do it thanks to the factory method... thanks for the insight, I will try it out and let you know.

I didn't use "cpu" time for naming since it would relate to what developers are used to know as cpu time (ie: actual time spent as active process and not in I/O wait), but thread time could be an option... I dont' know

Reply:
Yeah I figured out the need for thread-local later on when I got to the nesting stuff.  No way around that.

I'm okay with the naming used in the original report on the PR.  There's nothing standard you can say here. It is always going to take some documentation/explanation for anyone running multi-threaded to really interpret the meaning of the timings. 👍